File: sdk/typescript/src/proofs/quantum.ts
/**
 * QuantumProof helpers — assemble lightweight proof references from circuit & results.
 *
 * This client-side module helps you:
 *  1) Digest a quantum circuit specification and its measured results
 *  2) Package a QuantumProof "reference" object that node verifiers can bind to
 *     provider attestations, trap-circuit outcomes, and QoS metrics
 *  3) Produce/consume a minimal canonical CBOR envelope for transport
 *
 * NOTE:
 *  - Consensus-grade verification (provider attestations, trap checks, QoS/SLA)
 *    happens in the node. This module only handles canonicalization and hashing.
 */

import { sha3_256 } from '../utils/hash'
import { bytesToHex, hexToBytes, isHex } from '../utils/bytes'
import { encodeCanonicalCBOR, decodeCBOR } from '../utils/cbor'

// ──────────────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────────────

/** Provider attestation bundle (e.g., QPU identity, environment), opaque here. */
export type QuantumAttestation = Record<string, unknown>

/** Trap-circuit metadata for redundancy/accuracy checks. */
export interface QuantumTrapInfo {
  /** Deterministic seed describing trap selection (0x-hex). */
  trapSeed?: string
  /** Fraction of traps answered correctly, in [0..1]. */
  trapsRatio?: number
  /** Digest (0x-hex) of a separate receipts set, if used. */
  receiptsDigest?: string
}

/** Optional QoS metrics. */
export interface QuantumQoS {
  latencyMs?: number
  availability?: number   // [0..1]
  successRate?: number    // [0..1]
}

/** Canonical proof body verifiers expect. */
export interface QuantumProofBody {
  circuitDigest: string   // 0x-hex sha3_256(canonical circuit)
  resultDigest: string    // 0x-hex sha3_256(canonical results)
  shots?: number          // number of shots used (if applicable)
  device?: string         // logical device/model id
  providerId?: string
  jobId?: string
  attestation: QuantumAttestation
  traps?: QuantumTrapInfo
  qos?: QuantumQoS
  // Future fields allowed:
  [k: string]: unknown
}

/** On-wire reference envelope. */
export interface QuantumProofRef {
  typeId: 'QuantumProof' | number
  body: QuantumProofBody
  /** Deterministic nullifier to prevent duplicate submissions. */
  nullifier: string        // 0x-hex
}

/** Inputs for building a proof reference. */
export interface BuildQuantumProofInput {
  circuit: Uint8Array | string | Record<string, unknown> | unknown[]
  result: Uint8Array | string | Record<string, unknown> | unknown[]
  shots?: number
  device?: string
  providerId?: string
  jobId?: string
  attestation: QuantumAttestation
  traps?: QuantumTrapInfo
  qos?: QuantumQoS
}

/** Shallow validation result. */
export interface ValidateResult {
  ok: boolean
  errors: string[]
  warnings: string[]
}

// ──────────────────────────────────────────────────────────────────────────────
// Public API
// ──────────────────────────────────────────────────────────────────────────────

/** Digest a quantum circuit specification (object/bytes/string → sha3_256). */
export function digestCircuit(circuit: BuildQuantumProofInput['circuit']): string {
  const bytes = toBytesCanonical(circuit)
  return '0x' + bytesToHex(sha3_256(bytes))
}

/** Digest quantum results (counts/bitstrings/bytes → sha3_256). */
export function digestResult(result: BuildQuantumProofInput['result']): string {
  const bytes = toBytesCanonical(result)
  return '0x' + bytesToHex(sha3_256(bytes))
}

/**
 * Build a QuantumProofRef from circuit/result and metadata.
 * Nullifier binds (providerId?|jobId?|device?|circuitDigest|resultDigest).
 */
export function buildQuantumProofRef(input: BuildQuantumProofInput): QuantumProofRef {
  const circuitDigest = digestCircuit(input.circuit)
  const resultDigest = digestResult(input.result)

  const body: QuantumProofBody = {
    circuitDigest,
    resultDigest,
    shots: input.shots,
    device: input.device,
    providerId: input.providerId,
    jobId: input.jobId,
    attestation: input.attestation,
    traps: normalizeTraps(input.traps),
    qos: normalizeQoS(input.qos)
  }
  const nullifier = computeNullifier(body)
  return { typeId: 'QuantumProof', body, nullifier }
}

/** Deterministic nullifier = H("quantum.nullifier.v1"|providerId?|jobId?|device?|circuitDigest|resultDigest). */
export function computeNullifier(body: Pick<QuantumProofBody, 'providerId' | 'jobId' | 'device' | 'circuitDigest' | 'resultDigest'>): string {
  const dom = text('quantum.nullifier.v1')
  const parts: Uint8Array[] = [dom]
  if (body.providerId) parts.push(text(body.providerId))
  if (body.jobId) parts.push(text(body.jobId))
  if (body.device) parts.push(text(body.device))
  parts.push(hexToBytes(normalizeHex(body.circuitDigest)))
  parts.push(hexToBytes(normalizeHex(body.resultDigest)))
  return '0x' + bytesToHex(sha3_256(concat(...parts)))
}

/** Client-side format/range validation. */
export function validateQuantumProofBody(body: QuantumProofBody): ValidateResult {
  const errors: string[] = []
  const warnings: string[] = []

  try { normalizeHex(body.circuitDigest) } catch { errors.push('circuitDigest must be 0x-hex sha3_256') }
  try { normalizeHex(body.resultDigest) } catch { errors.push('resultDigest must be 0x-hex sha3_256') }

  if (body.shots != null) {
    if (!Number.isFinite(body.shots) || body.shots <= 0) errors.push('shots must be a positive number')
  }
  if (!body.attestation || typeof body.attestation !== 'object') errors.push('attestation missing/invalid')

  if (body.traps?.trapsRatio != null) {
    if (!(body.traps.trapsRatio >= 0 && body.traps.trapsRatio <= 1)) errors.push('traps.trapsRatio must be in [0,1]')
  }
  if (body.qos?.availability != null && !(body.qos.availability >= 0 && body.qos.availability <= 1)) {
    errors.push('qos.availability must be in [0,1]')
  }
  if (body.qos?.successRate != null && !(body.qos.successRate >= 0 && body.qos.successRate <= 1)) {
    errors.push('qos.successRate must be in [0,1]')
  }

  return { ok: errors.length === 0, errors, warnings }
}

/** Encode a QuantumProofRef into canonical CBOR. */
export function toEnvelopeCBOR(ref: QuantumProofRef): Uint8Array {
  const env: QuantumProofRef = {
    typeId: typeof ref.typeId === 'number' ? ref.typeId : 'QuantumProof',
    body: {
      ...ref.body,
      circuitDigest: normalizeHex(ref.body.circuitDigest),
      resultDigest: normalizeHex(ref.body.resultDigest),
      traps: normalizeTraps(ref.body.traps),
      qos: normalizeQoS(ref.body.qos)
    },
    nullifier: normalizeHex(ref.nullifier)
  }
  return encodeCanonicalCBOR(env)
}

/** Decode a canonical CBOR envelope into a normalized QuantumProofRef. */
export function fromEnvelopeCBOR(data: Uint8Array): QuantumProofRef {
  const env = decodeCBOR(data) as QuantumProofRef
  if (!env || typeof env !== 'object') throw new Error('invalid envelope')
  if (!('typeId' in env) || !('body' in env) || !('nullifier' in env)) throw new Error('missing envelope fields')
  const out: QuantumProofRef = {
    typeId: (env.typeId === 'QuantumProof' || typeof env.typeId === 'number') ? env.typeId : 'QuantumProof',
    body: {
      ...env.body,
      circuitDigest: normalizeHex(env.body.circuitDigest),
      resultDigest: normalizeHex(env.body.resultDigest),
      traps: normalizeTraps(env.body.traps),
      qos: normalizeQoS(env.body.qos)
    },
    nullifier: normalizeHex(env.nullifier)
  }
  return out
}

// ──────────────────────────────────────────────────────────────────────────────
/* Internals */
// ──────────────────────────────────────────────────────────────────────────────

function normalizeHex(h: string): string {
  if (typeof h !== 'string') throw new Error('expected hex string')
  const s = h.startsWith('0x') || h.startsWith('0X') ? h : ('0x' + h)
  if (!isHex(s)) throw new Error(`invalid hex: ${h}`)
  return s.toLowerCase()
}

function text(s: string): Uint8Array {
  return new TextEncoder().encode(s)
}

function concat(...parts: Uint8Array[]): Uint8Array {
  if (parts.length === 0) return new Uint8Array(0)
  const n = parts.reduce((a, p) => a + p.length, 0)
  const out = new Uint8Array(n)
  let off = 0
  for (const p of parts) { out.set(p, off); off += p.length }
  return out
}

/** Normalize traps object & basic hex checks. */
function normalizeTraps(t?: QuantumTrapInfo): QuantumTrapInfo | undefined {
  if (!t) return undefined
  const traps: QuantumTrapInfo = { ...t }
  if (traps.trapSeed != null) traps.trapSeed = normalizeHex(traps.trapSeed)
  if (traps.receiptsDigest != null) traps.receiptsDigest = normalizeHex(traps.receiptsDigest)
  return traps
}

/** Normalize QoS fields to numbers (drop NaN/undefined). */
function normalizeQoS(q?: QuantumQoS): QuantumQoS | undefined {
  if (!q) return undefined
  const out: QuantumQoS = {}
  if (q.latencyMs != null) out.latencyMs = Number(q.latencyMs)
  if (q.availability != null) out.availability = Number(q.availability)
  if (q.successRate != null) out.successRate = Number(q.successRate)
  return out
}

/** Deterministic bytes for arbitrary inputs (Uint8Array | hex | utf8 | object). */
function toBytesCanonical(x: Uint8Array | string | Record<string, unknown> | unknown[]): Uint8Array {
  if (x instanceof Uint8Array) return x
  if (typeof x === 'string') {
    if (isHex(x)) return hexToBytes(x)
    return new TextEncoder().encode(x)
  }
  return encodeCanonicalCBOR(x as any)
}

// ──────────────────────────────────────────────────────────────────────────────
// Default export
// ──────────────────────────────────────────────────────────────────────────────

export default {
  digestCircuit,
  digestResult,
  buildQuantumProofRef,
  computeNullifier,
  validateQuantumProofBody,
  toEnvelopeCBOR,
  fromEnvelopeCBOR
}

File: sdk/rust/src/proofs/quantum.rs
//! QuantumProof assembly helpers (developer-facing).
//!
//! This module helps SDK users assemble a **Quantum proof reference**
//! from the quantum circuit description, provider identity/attestation,
//! trap-circuit outcomes, and optional QoS metrics. It mirrors the
//! shape expected by node tooling that serializes canonical envelopes
//! in the `proofs/` module, but keeps things lightweight for clients.
//
// Design notes
// - We bind the circuit via `circuit_digest = sha3_256(circuit_bytes)`.
// - Optionally bind an output digest if your flow stores the raw outputs.
// - We expose a small "bench" estimator to compute `quantum_units` from
//   (depth × width × shots); networks may tune coefficients on-chain.
// - We validate obvious bounds (ratios in [0,1], shots > 0, hex lengths).
//
// This is **not** a consensus implementation. It produces a reference
// object that higher-level tooling can convert into the canonical CBOR
// envelope expected by validators.

use crate::error::{Error, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use sha3::{Digest, Sha3_256};
use std::collections::BTreeMap;

/// Provider identity/attestation bundle (opaque to the SDK).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuantumProviderCert {
    /// Human/registry identifier, e.g. "qpu:acme-iontrap-1" or "sim:oqc".
    pub provider: String,
    /// Evidence object (X.509 chain, EdDSA/PQ hybrid, statements, etc.).
    pub evidence: JsonValue,
    /// Future-proof fields.
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// Trap-circuit outcomes (for correctness confidence).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuantumTraps {
    pub passed: u32,
    pub total: u32,
    /// Optional ratio; computed when absent on `validate()`.
    #[serde(default)]
    pub ratio: Option<f64>,
    /// Optional Wilson/Clopper-Pearson lower bound in [0,1].
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

impl QuantumTraps {
    pub fn compute_ratio(&self) -> f64 {
        if self.total == 0 {
            0.0
        } else {
            (self.passed as f64) / (self.total as f64)
        }
    }
}

/// QoS metrics for the quantum provider/job.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct QpuQos {
    #[serde(default)]
    pub latency_ms: Option<u64>,
    #[serde(default)]
    pub availability: Option<f64>,
    #[serde(default)]
    pub success: Option<bool>,
    #[serde(default)]
    pub score: Option<f64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// Optional complexity meta for pricing/visualization.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CircuitComplexity {
    #[serde(default)]
    pub depth: Option<u32>,
    #[serde(default)]
    pub width: Option<u32>,
    #[serde(default)]
    pub shots: Option<u32>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// A reference object your app can pass to proof builders on the node.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuantumProofRef {
    /// SHA3-256 digest of the circuit description (0x-hex, 32 bytes).
    pub circuit_digest: String,
    /// Number of shots requested/executed for this run.
    pub shots: u32,
    /// Optional SHA3-256 digest of the raw outputs (0x-hex, 32 bytes).
    #[serde(default)]
    pub output_digest: Option<String>,
    /// Provider certificate/attestation.
    pub provider_cert: QuantumProviderCert,
    /// Optional trap-circuit outcomes.
    #[serde(default)]
    pub traps: Option<QuantumTraps>,
    /// Optional QoS metrics.
    #[serde(default)]
    pub qos: Option<QpuQos>,
    /// Optional complexity meta (depth/width/shots).
    #[serde(default)]
    pub complexity: Option<CircuitComplexity>,
    /// Optional computed pricing/work units (network-specific).
    #[serde(default)]
    pub quantum_units: Option<u64>,
    /// Free-form application metadata.
    #[serde(default)]
    pub meta: BTreeMap<String, JsonValue>,
    /// Future-proof fields.
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

impl QuantumProofRef {
    /// Construct from circuit bytes and provider certificate.
    pub fn from_circuit(
        circuit_bytes: &[u8],
        shots: u32,
        provider_cert: QuantumProviderCert,
        traps: Option<QuantumTraps>,
        qos: Option<QpuQos>,
        complexity: Option<CircuitComplexity>,
    ) -> Self {
        let mut hasher = Sha3_256::new();
        hasher.update(circuit_bytes);
        let digest = hasher.finalize();

        Self {
            circuit_digest: hex0x(&digest),
            shots,
            output_digest: None,
            provider_cert,
            traps,
            qos,
            complexity,
            quantum_units: None,
            meta: BTreeMap::new(),
            extra: BTreeMap::new(),
        }
    }

    /// Bind an output digest (e.g., result payload), if available.
    pub fn with_output_digest(mut self, output_bytes: &[u8]) -> Self {
        let mut h = Sha3_256::new();
        h.update(output_bytes);
        let d = h.finalize();
        self.output_digest = Some(hex0x(&d));
        self
    }

    /// Attach a computed `quantum_units` value.
    pub fn with_units(mut self, units: u64) -> Self {
        self.quantum_units = Some(units);
        self
    }

    /// Validate invariants and fill derived fields where applicable.
    pub fn validate(&mut self) -> Result<()> {
        if self.shots == 0 {
            return Err(Error::InvalidData("shots must be > 0".into()));
        }
        if !is_hex32(&self.circuit_digest) {
            return Err(Error::InvalidData("circuitDigest must be 0x + 64 hex chars".into()));
        }
        if let Some(od) = self.output_digest.as_ref() {
            if !is_hex32(od) {
                return Err(Error::InvalidData("outputDigest must be 0x + 64 hex chars".into()));
            }
        }
        if let Some(t) = self.traps.as_mut() {
            let ratio = t.ratio.unwrap_or_else(|| t.compute_ratio());
            if !(0.0..=1.0).contains(&ratio) {
                return Err(Error::InvalidData(format!("trap ratio out of range: {ratio}")));
            }
            if let Some(c) = t.confidence {
                if !(0.0..=1.0).contains(&c) {
                    return Err(Error::InvalidData(format!("confidence out of range: {c}")));
                }
            }
            t.ratio = Some(ratio);
        }
        if let Some(q) = self.qos.as_ref() {
            if let Some(a) = q.availability {
                if !(0.0..=1.0).contains(&a) {
                    return Err(Error::InvalidData(format!("availability out of range: {a}")));
                }
            }
            if let Some(s) = q.score {
                if !(0.0..=1.0).contains(&s) {
                    return Err(Error::InvalidData(format!("score out of range: {s}")));
                }
            }
        }
        if let Some(c) = self.complexity.as_ref() {
            if let Some(w) = c.width {
                if w == 0 {
                    return Err(Error::InvalidData("width must be > 0".into()));
                }
            }
            if let Some(d) = c.depth {
                if d == 0 {
                    return Err(Error::InvalidData("depth must be > 0".into()));
                }
            }
            if let Some(s) = c.shots {
                if s == 0 {
                    return Err(Error::InvalidData("complexity.shots must be > 0".into()));
                }
            }
        }
        Ok(())
    }
}

/// Coefficients for unit estimation from (depth, width, shots).
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BenchCoeffs {
    /// Additive base per shot.
    pub base_per_shot: f64,
    /// Linear depth contribution per shot.
    pub depth_per_shot: f64,
    /// Linear width contribution per shot.
    pub width_per_shot: f64,
    /// Optional multiplicative scale (e.g., to convert to "units").
    pub scale: f64,
}

impl Default for BenchCoeffs {
    fn default() -> Self {
        Self {
            base_per_shot: 1.0,
            depth_per_shot: 0.25,
            width_per_shot: 0.50,
            scale: 1.0,
        }
    }
}

/// Estimate quantum work units.
/// This is a simple linear model intended for UI/pricing hints.
///
/// units = scale * shots * (base + depth*α + width*β)
pub fn estimate_quantum_units(depth: u32, width: u32, shots: u32, coeffs: BenchCoeffs) -> u64 {
    let d = depth as f64;
    let w = width as f64;
    let s = shots as f64;
    let per_shot = coeffs.base_per_shot + d * coeffs.depth_per_shot + w * coeffs.width_per_shot;
    let units = coeffs.scale * s * per_shot;
    if units.is_finite() && units >= 0.0 {
        units.round() as u64
    } else {
        0
    }
}

// ------------------------------ Utilities -----------------------------------

fn hex0x(bytes: impl AsRef<[u8]>) -> String {
    format!("0x{}", hex::encode(bytes))
}

fn is_hex32(s: &str) -> bool {
    s.starts_with("0x") && s.len() == 66 && s[2..].bytes().all(|b| matches!(b, b'0'..=b'9'|b'a'..=b'f'|b'A'..=b'F'))
}

// ---------------------------------- Tests -----------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn cert() -> QuantumProviderCert {
        QuantumProviderCert {
            provider: "qpu:acme-iontrap-1".into(),
            evidence: json!({
                "certChain": ["-----BEGIN CERT-----..."],
                "pubkey": "ed25519:abcd..."
            }),
            extra: BTreeMap::new(),
        }
    }

    #[test]
    fn build_validate_and_units() {
        let circ = br#"{"qasm":"OPENQASM 3; qubit[4] q; h q[0]; cx q[0], q[1];"}"#;
        let complexity = CircuitComplexity {
            depth: Some(32),
            width: Some(4),
            shots: Some(128),
            extra: BTreeMap::new(),
        };
        let mut pref = QuantumProofRef::from_circuit(
            circ,
            128,
            cert(),
            Some(QuantumTraps {
                passed: 95,
                total: 100,
                ratio: None,
                confidence: Some(0.95),
                extra: BTreeMap::new(),
            }),
            Some(QpuQos {
                latency_ms: Some(2500),
                availability: Some(0.999),
                success: Some(true),
                score: Some(0.97),
                extra: BTreeMap::new(),
            }),
            Some(complexity),
        );
        let units = estimate_quantum_units(32, 4, 128, BenchCoeffs::default());
        pref = pref.with_units(units);
        pref.validate().unwrap();
        assert!(pref.circuit_digest.starts_with("0x"));
        assert_eq!(pref.shots, 128);
        assert_eq!(pref.quantum_units, Some(units));
        assert!(pref.traps.as_ref().unwrap().ratio.unwrap() > 0.0);
    }

    #[test]
    fn output_digest_attachment() {
        let circ = b"circuit";
        let output = b"measurement-bits";
        let mut pref = QuantumProofRef::from_circuit(circ, 10, cert(), None, None, None)
            .with_output_digest(output);
        pref.validate().unwrap();
        assert!(pref.output_digest.as_ref().unwrap().starts_with("0x"));
    }

    #[test]
    fn bad_bounds_fail() {
        let circ = b"x";
        let mut pref = QuantumProofRef::from_circuit(
            circ,
            0, // invalid shots
            cert(),
            None,
            Some(QpuQos {
                availability: Some(1.5),
                ..Default::default()
            }),
            None,
        );
        assert!(pref.validate().is_err());
    }
}
